pkg/ddl/ingest/engine_mgr.go

```diff
@@ -15,118 +15,109 @@
 		}
-		en.writerCount++
-		info = LitInfoAddWriter
+
+		openedEngines[indexID] = newEngineInfo(
+			bc.ctx,
+			bc.jobID,
+			indexID,
+			cfg,
+			bc.cfg,
+			openedEngine,
+			openedEngine.GetEngineUUID(),
+			bc.memRoot,
+		)
 	}
-	bc.MemRoot.ConsumeWithTag(encodeEngineTag(jobID, indexID), int64(bc.cfg.TikvImporter.LocalWriterMemCacheSize))
-	logutil.Logger(bc.ctx).Info(info, zap.Int64("job ID", jobID),
-		zap.Int64("index ID", indexID),
-		zap.Int64("current memory usage", bc.MemRoot.CurrentUsage()),
-		zap.Int64("memory limitation", bc.MemRoot.MaxMemoryQuota()),
-		zap.Int("current writer count", en.writerCount))
-	return en, nil
-}
 
-// Unregister delete the engineInfo from the engineManager.
-func (bc *litBackendCtx) Unregister(jobID, indexID int64) {
-	ei, exist := bc.Load(indexID)
-	if !exist {
-		return
+	for _, indexID := range indexIDs {
+		ei := openedEngines[indexID]
+		ret = append(ret, ei)
+		bc.engines[indexID] = ei
 	}
+	bc.memRoot.Consume(numIdx * (structSizeEngineInfo + engineCacheSize))
 
-	ei.Clean()
-	bc.Delete(indexID)
-	bc.MemRoot.ReleaseWithTag(encodeEngineTag(jobID, indexID))
-	bc.MemRoot.Release(StructSizeWriterCtx * int64(ei.writerCount))
-	bc.MemRoot.Release(StructSizeEngineInfo)
+	logutil.Logger(bc.ctx).Info(LitInfoOpenEngine, zap.Int64("job ID", bc.jobID),
+		zap.Int64s("index IDs", indexIDs),
+		zap.Int64("current memory usage", bc.memRoot.CurrentUsage()),
+		zap.Int64("memory limitation", bc.memRoot.MaxMemoryQuota()))
+	return ret, nil
 }
 
-// ResetWorkers reset the writer count of the engineInfo because
-// the goroutines of backfill workers have been terminated.
-func (bc *litBackendCtx) ResetWorkers(jobID int64) {
-	for _, indexID := range bc.Keys() {
-		ei, exist := bc.Load(indexID)
-		if !exist {
-			continue
-		}
-		bc.MemRoot.Release(StructSizeWriterCtx * int64(ei.writerCount))
-		bc.MemRoot.ReleaseWithTag(encodeEngineTag(jobID, indexID))
-		engineCacheSize := int64(bc.cfg.TikvImporter.EngineMemCacheSize)
-		bc.MemRoot.ConsumeWithTag(encodeEngineTag(jobID, indexID), engineCacheSize)
-		ei.writerCount = 0
+// UnregisterEngines implements BackendCtx.
+func (bc *litBackendCtx) UnregisterEngines() {
+	numIdx := int64(len(bc.engines))
+	for _, ei := range bc.engines {
+		ei.Clean()
 	}
-}
+	bc.engines = make(map[int64]*engineInfo, 10)
 
-// unregisterAll delete all engineInfo from the engineManager.
-func (bc *litBackendCtx) unregisterAll(jobID int64) {
-	for _, idxID := range bc.Keys() {
-		bc.Unregister(jobID, idxID)
-	}
+	engineCacheSize := int64(bc.cfg.TikvImporter.EngineMemCacheSize)
+	bc.memRoot.Release(numIdx * (structSizeEngineInfo + engineCacheSize))
 }
 
-func encodeEngineTag(jobID, indexID int64) string {
-	return fmt.Sprintf("%d-%d", jobID, indexID)
+// FinishedWritingNeedImport implements BackendCtx.
+func (bc *litBackendCtx) FinishedWritingNeedImport() bool {
+	if len(bc.engines) == 0 {
+		return false
+	}
+	for _, ei := range bc.engines {
+		return ei.closedEngine != nil
```

Comment:
What if some of the engines are closed, others are not?

Reply:
https://github.com/pingcap/tidb/pull/53233/commits/83b8ab994abff1781190d678c6ebefe23f9b7e39